### PR TITLE
Make DiffusionGemma trainable

### PR DESCRIPTION
**Pull request by kashif**
Three small fixes so DiffusionGemma works with Trainer and SFT recipes.

`bos_token_id` was missing from the strict generation config, which crashed `align_special_tokens` and broke checkpoint reload. The forward now also returns the encoder's last hidden state (needed for the encoder AR co-loss used in the reference fine-tuning recipe) and accepts a per-example `self_conditioning_mask`. Gradient checkpointing is declared unsupported since checkpointed layers drop the encoder KV cache writes the decoder reads from, which silently broke training.

Inference is unchanged, all diffusion_gemma tests pass.

cc @gante

### CODE REVIEW

**Comment by gante** [comment]
The changes LGTM 👍

File: src/transformers/models/diffusion_gemma/modeling_diffusion_gemma.py [hunk 1]
```diff
@@ -1536,16 +1579,18 @@ def forward(
...
             **kwargs,
         )
 
-        return BaseModelOutputWithPast(
+        return DiffusionGemmaModelOutputWithPast(
             last_hidden_state=decoder_outputs.last_hidden_state,
             hidden_states=decoder_outputs.hidden_states,
             attentions=decoder_outputs.attentions,
             past_key_values=past_key_values,
+            encoder_last_hidden_state=encoder_last_hidden_state,
```
Comment (zucchini-nlp) [inline src/transformers/models/diffusion_gemma/modeling_diffusion_gemma.py:1593]: nit: we could use a `Seq2SeqModelOutputWithPast` imo

**Review by zucchini-nlp — APPROVED** [review]
Yep, lgtm. Maybe could use an exitsing model output class but otherwise fine

**Comment by kashif** [comment]
Good point, switched to Seq2SeqModelOutput / Seq2SeqLMOutput and dropped the custom classes.

**Comment by kashif** [comment]
Reverted the Seq2Seq output switch after a closer look: it renames hidden_states and attentions to decoder_*, which breaks anyone already reading the 5.11 outputs. The model specific subclasses keep the old field names and just add encoder_last_hidden_state.
